randovania/generator/dock_weakness_distributor.py

```diff
@@ -0,0 +1,171 @@
+
+    for player, patches in all_patches.items():
+        if patches.configuration.dock_rando.mode != DockRandoMode.VANILLA:
+            unassigned_docks.extend(((player, identifier) for identifier in patches.dock_weakness.keys()))
+    
+    rng.shuffle(unassigned_docks)
+
+    docks_placed = 0
+    docks_to_place = len(unassigned_docks)
+
+    while unassigned_docks:
+        await asyncio.sleep(0)
+        status_update(f"{docks_placed}/{docks_to_place} docks placed")
+
+        player, identifier = unassigned_docks.pop()
+
+        game = descriptions[player]
+        patches = all_patches[player]
+
+        dock: DockNode = game.world_list.node_by_identifier(identifier)
+        target: DockNode = game.world_list.node_by_identifier(dock.get_target_identifier(NodeContext(
+            patches,
+            patches.starting_items,
+            game.resource_database,
+            game.world_list,
+        )))
+
+        dock_type_params = game.dock_weakness_database.dock_rando_params[dock.dock_type]
+        dock_type_state = patches.configuration.dock_rando.types_state[dock.dock_type]
+
+        # Calculate the reach assuming this dock is completely locked
+        locks = {identifier: dock_type_params.locked}
+        if patches.configuration.dock_rando.mode == DockRandoMode.TWO_WAY:
+            locks[target.identifier] = dock_type_params.locked
+        locked_patches = patches.assign_dock_weakness_assignment(locks)
+
+        state, logic = resolver.setup_resolver(patches.configuration, locked_patches)
+        logic = DockRandoLogic.from_logic(logic, dock)
+        debug.log_resolve_start()
+        debug.debug_print(f"Starting {identifier} at {time.perf_counter()}")
+        
+        level = debug.debug_level()
+        debug.set_level(0)
+
+        state_async = resolver.advance_depth(state, logic, lambda s: None)
+        try:
+            new_state = await asyncio.wait_for(state_async, RESOLVER_TIMEOUT)
+        except asyncio.TimeoutError:
+            debug.set_level(level)
+            new_state = None
+            debug.debug_print("Timeout")
+        else:
+            debug.set_level(level)
+            debug.debug_print("Finished resolver")
+
+        # Determine valid weaknesses to assign
+        possible_weaknesses = [dock_type_params.unlocked]
+
+        if new_state is not None:
+            reach = ResolverReach.calculate_reach(logic, new_state)
+
+            if dock_type_params.locked in dock_type_state.can_change_to and target in reach.nodes:
+                possible_weaknesses.append(dock_type_params.locked)
+            
+            debug.debug_print(f"{[res.long_name for res in new_state.resources if isinstance(res, (ItemResourceInfo, SimpleResourceInfo)) and res.resource_type in {ResourceType.ITEM, ResourceType.EVENT}]}")
+            possible_weaknesses.extend((
+                weakness for weakness in dock_type_state.can_change_to.difference(possible_weaknesses) if (
```

Comment:
Iterating over `set`. That is highly non-deterministic and should be avoided in anything generator related.